src/bootstrap/native.rs

```diff
@@ -916,7 +916,51 @@ impl Step for Lld {
         t!(fs::create_dir_all(&out_dir));
 
         let mut cfg = cmake::Config::new(builder.src.join("src/llvm-project/lld"));
-        configure_cmake(builder, target, &mut cfg, true, LdFlags::default());
+        let mut ldflags = LdFlags::default();
+
+        // When building LLD as part of a build that asks LLVM to generate profiles, for example
+        // when doing PGO on CI, something on windows (maybe `clang-cl`) doesn't automatically link
+        // the profiler runtime in. We manually link the lib in that case to avoid linking errors:
+        // it's also located in the same precompiled LLVM distribution as `clang-cl`.
+        //
+        // Since we only do PGO on CI for x64 windows, handle that specifically, only for the config
+        // the bots use on CI (like we do for building llvm itself).
+        if builder.config.llvm_profile_generate
+            && target.contains("msvc")
+            && target.contains("x86_64")
+            && builder.ninja()
+            && builder.config.ccache.is_some()
+        {
+            // `clang-cl` is itself located in the `$LLVM_DISTRO_ROOT/bin` subfolder, let's use it
+            // to find the root.
+            let clang_cl_path = Path::new(
+                builder.config.llvm_clang_cl.as_ref().expect("`llvm.clang-cl` should be set here"),
+            );
+            let llvm_root = clang_cl_path
+                .parent()
+                .expect("The LLVM bin root should be the parent of `clang-cl`")
+                .parent()
+                .expect("The LLVM root should be the parent of `bin`");
+
+            // The profiler runtime is itself in `lib/clang/$LLVM_VERSION/lib/windows`.
+            let clang_libs = llvm_root.join("lib").join("clang");
+
+            // There should only be one version folder here, for the given distribution, except with
+            // broken LLVM releases like 14.0.2.
+            let mut entries = fs::read_dir(clang_libs).expect("The clang lib folder should exist");
+            let llvm_version = entries
+                .next()
+                .expect("reading the clang entries")
+                .expect("reading the 1st clang entry")
+                .path();
+            let profile_rt =
+                llvm_version.join("lib").join("windows").join("clang_rt.profile-x86_64.lib");
+
+            // Push it to the cmake linker flags.
+            ldflags.push_all(&profile_rt);
+        }
```

Comment:
It looks like LLVM's CMAKE setup does something similar here too:
https://github.com/llvm/llvm-project/blob/782e91224601e461c019e0a4573bbccc6094fbcd/llvm/cmake/modules/HandleLLVMOptions.cmake#L1058-L1079

It uses `clang-cl /clang:-print-libgcc-file-name /clang:--rtlib=compiler-rt` to make Clang print the path to the compiler-rt lib. That should be more robust than the current code.

Reply:
I've done that for LLD/LLVM, it's cleaner but still requires the rustflags link-arg workaround for now unfortunately, as mentioned below.